proposer_stakable_balance.py is floating

Flaked on travis in #908 but does not seem to be connected to the branch contents:
```
2019-04-11 07:50:12.622000 TestFramework (INFO): Waiting for nodes to have started up...
2019-04-11 07:50:12.622000 TestFramework (INFO): Checking that no node is proposing as no node has a peer right now
2019-04-11 07:50:12.633000 TestFramework (INFO): Connecting nodes
2019-04-11 07:50:13.868000 TestFramework (INFO): Waiting for nodes to be connected (should read NOT_PROPOSING_NOT_ENOUGH_BALANCE then)
2019-04-11 07:50:14.194000 TestFramework (INFO): The nodes with funds should advance to IS_PROPOSING
2019-04-11 07:50:14.217000 TestFramework (INFO): The others should stay in NOT_ENOUGH_BALANCE
2019-04-11 07:50:14.228000 TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/travis/build/dtr-org/unit-e/build/unit-e-i686-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 156, in main
    self.run_test()
  File "/home/travis/build/dtr-org/unit-e/build/unit-e-i686-pc-linux-gnu/test/functional/proposer_stakeable_balance.py", line 83, in run_test
    assert_equal(wallet['stakeable_balance'], Decimal('10000.00000000'))
  File "/home/travis/build/dtr-org/unit-e/build/unit-e-i686-pc-linux-gnu/test/functional/test_framework/util.py", line 41, in assert_equal
    raise AssertionError("not(%s)" % " == ".join(str(arg) for arg in (thing1, thing2) + args))
AssertionError: not(0E-8 == 10000.00000000)
```
[full log](https://github.com/dtr-org/unit-e/files/3067652/proposer_stakeable_balance.log)

--- a/src/proposer/proposer.cpp
+++ b/src/proposer/proposer.cpp
@@ -102,6 +102,8 @@ class ProposerImpl : public Proposer {
         }
         std::shared_ptr<const CBlock> block;
         {
+          // To pick up to date coins for staking we need to make sure that the wallet is synced to the current chain.
+          wallet->BlockUntilSyncedToCurrentChain();
           LOCK2(m_active_chain->GetLock(), wallet_ext.GetLock());
           const CBlockIndex &tip = *m_active_chain->GetTip();
           const staking::CoinSet coins = wallet_ext.GetStakeableCoins();

--- a/src/proposer/proposer_rpc.cpp
+++ b/src/proposer/proposer_rpc.cpp
@@ -30,9 +30,10 @@ class ProposerRPCImpl : public ProposerRPC {
   const Dependency<Proposer> m_proposer;
 
   UniValue GetWalletInfo(const std::vector<CWalletRef> &wallets) const {
-    LOCK(m_chain->GetLock());
     UniValue result(UniValue::VARR);
     for (const auto &wallet : wallets) {
+      wallet->BlockUntilSyncedToCurrentChain();
+      LOCK(m_chain->GetLock());
       const auto &wallet_extension = wallet->GetWalletExtension();
       const auto &proposerState = wallet_extension.GetProposerState();
       UniValue info(UniValue::VOBJ);
@@ -145,6 +146,7 @@ class ProposerRPCImpl : public ProposerRPC {
           "get the stakeable coins\n",
           __func__));
     }
+    wallet->BlockUntilSyncedToCurrentChain();
     UniValue obj(UniValue::VOBJ);
     const staking::StakingWallet &staking_wallet = wallet->GetWalletExtension();
     const staking::CoinSet stakeable_coins = [&]() {

--- a/src/staking/legacy_validation_interface.h
+++ b/src/staking/legacy_validation_interface.h
@@ -39,7 +39,7 @@ class StakeValidator;
 //!
 //! For backwards compatibility CValidationState has been augmented
 //! with staking::BlockValidationInfo. The scope of it is narrower
-//! then for CValidationState as we also have BlockValidationResult.
+//! than for CValidationState as we also have BlockValidationResult.
 //! To translate into the existing validation interface different
 //! implementations are provided for the LegacyValidationInterface:
 //! - LegacyValidationInterface::New which accesses the PoS BlockValidator

--- a/src/wallet/rpcwallet.cpp
+++ b/src/wallet/rpcwallet.cpp
@@ -3442,6 +3442,11 @@ UniValue generateBlocks(CWallet * const pwallet, std::shared_ptr<CReserveScript>
     }
     unsigned int nExtraNonce = 0;
     UniValue blockHashes(UniValue::VARR);
+
+    // To pick up to date coins for staking we need to make sure that the wallet is synced to the current chain.
+    if (pwallet) {
+        pwallet->BlockUntilSyncedToCurrentChain();
+    }
     while (nHeight < nHeightEnd)
     {
         std::unique_ptr<CBlockTemplate> pblocktemplate(BlockAssembler(Params()).CreateNewBlock(

--- a/test/functional/proposer_stakeable_balance.py
+++ b/test/functional/proposer_stakeable_balance.py
@@ -81,7 +81,7 @@ class ProposerStakeableBalanceTest(UnitETestFramework):
             wallet = status['wallets'][0]
             assert_equal(wallet['balance'], Decimal('10000.00000000'))
             assert_equal(wallet['stakeable_balance'], Decimal('10000.00000000'))
-        # and others shoulds till not have enough funds for proposing blocks
+        # and others should still not have enough funds for proposing blocks
         for i in range(num_keys, self.num_nodes):
             status = nodes[i].proposerstatus()
             wallet = status['wallets'][0]

